Add pyHIPIFY scripts needed for ROCm transpilation to PyTorch

iotamudelta opened the pull request:
As discussed in call, this will allow us to keep this integral part of the effort to run PyTorch on ROCm in sync with the main code.

iotamudelta committed:
Merge pull request #1 from iotamudelta/master

Bring our fork up to date.

ezyang committed:
Empty commit to test CI.

iotamudelta committed:
Merge remote-tracking branch 'upstream/master'

iotamudelta committed:
Merge pull request #6 from iotamudelta/master

Merge from upstream master into our fork.

iotamudelta committed:
Merge remote-tracking branch 'upstream/master'

iotamudelta committed:
As per call - add pyHIPIFY directly to github.

iotamudelta committed:
Merge pull request #8 from iotamudelta/master

sync w/ upstream

iotamudelta committed:
Merge remote-tracking branch 'rocm_upstream/master' into include_pyhipigy

iotamudelta committed:
Merge pull request #9 from iotamudelta/include_pyhipigy

add pyHIPIFY to PyTorch repo

bddppq reviewed (commented):
Thanks for merging these python scripts into pytorch repo.
The CI failure is python code [lint check](https://travis-ci.org/pytorch/pytorch/jobs/395669595), you can use `autopep8` to automatically make them conforming the style requirements.

bddppq commented on tools/amd_build/pyHIPIFY/hipify-python.py:
```diff
@@ -0,0 +1,946 @@
+#!/usr/bin/python
+""" The Python Hipify script.
+##
+# Copyright (c) 2015-2016 Advanced Micro Devices, Inc. All rights reserved.
```
@soumith re. copyright

soumith replied:
once we make modifications (going forward), we can add a copyright line that says "portions copyright AMD, portions copyright Facebook"

iotamudelta replied:
updated copyright, now also reflects correctly that both AMD and Facebook have contributed already to pyHIPIFY

soumith commented:
the script is failing lint

ezyang commented on tools/amd_build/pyHIPIFY/cuda_to_hip_mappings.py:
```diff
@@ -0,0 +1,2097 @@
+from constants import *
+
+CUDA_TYPE_NAME_MAP = {
```
This file needs a description of what the format of the entries of the map are. I see some entries which are 3-tuples, and some which are 4-tuples. What are the valid values at each tuple?

ezyang replied:
Still unaddressed

iotamudelta replied:
done now - also referencing the hipify-clang source this came from

ezyang commented on tools/amd_build/pyHIPIFY/cuda_to_hip_mappings.py:
```diff
@@ -0,0 +1,2097 @@
+     "CU_RES_VIEW_FORMAT_FLOAT_2X16": ("hipResViewFormatHalf2", CONV_TEX,API_DRIVER),
+     "CU_RES_VIEW_FORMAT_FLOAT_4X16": ("hipResViewFormatHalf4", CONV_TEX,API_DRIVER),
+     "CU_RES_VIEW_FORMAT_FLOAT_1X32": ("hipResViewFormatFloat1", CONV_TEX,API_DRIVER),
+     "CU_RES_VIEW_FORMAT_FLOAT_2X32": ("hipResVi
```
Can we sort these before the initial commit? Will make it easier to find things / make modifications later.

ezyang commented on tools/amd_build/pyHIPIFY/hipify-python.py:
```diff
@@ -0,0 +1,946 @@
+
+                # Update the progress
+                print(os.path.join(dirpath, filename))
+                update_progress_bar(total_files, current_file)
```
You don't have to fix this now, but for future reference, whenever you add some sort of progress bar functionality, always include an option to disable it. Otherwise, these progress bars clog up CI logs without adding much of value.

iotamudelta replied:
agreed.

Jorghi12 replied:
This should be fixed now. A simple addition to the command line invocation for hipify-python is needed.

ezyang commented on tools/amd_build/pyHIPIFY/hipify-python.py:
```diff
@@ -0,0 +1,946 @@
+
+def openf(filename, mode):
+   if sys.version_info[0] == 3:
+      return open(filename, mode, errors='ignore')
```
This should not be necessary. All of the files in our repository should be UTF-8 encoded. Is this not true?

ezyang commented on tools/amd_build/pyHIPIFY/hipify-python.py:
```diff
@@ -0,0 +1,946 @@
+        status)
+
+    # Send the progress to stdout.
+    sys.stdout.write(text)
```
More standard practice is to send this kind of info to stderr. It's more conventional, because it keeps stdout clutter free for the case that there is actually important data you want to pipe out from it (in this case, there isn't).

iotamudelta replied:
done.

ezyang commented on tools/amd_build/pyHIPIFY/hipify-python.py:
```diff
@@ -0,0 +1,946 @@
+
+def filename_ends_with_extension(filename, extensions):
+    """Helper method to see if filename ends with certain extension"""
+    return reduce(lambda result, ext: filename.endswith("." + ext) or result, extensions, False)
```
Oh man, Guido would totally be rolling in his... bed... at this code. ;)

iotamudelta replied:
leave or fix this now?

ezyang replied:
fine for now

Jorghi12 replied:
Do you think we could get code reviews from Guido?

ezyang commented on tools/amd_build/pyHIPIFY/hipify-python.py:
```diff
@@ -0,0 +1,946 @@
+def processKernelLaunches(string, stats):
+    """ Replace the CUDA style Kernel launches with the HIP style kernel launches."""
+    # Concat the namespace with the kernel names. (Find cleaner way of doing this later).
+    string = re.sub(r'([ ]+)(detail+)::[ ]+\\\n[ ]+', lambda inp: "%s%s::" % (inp.group(1), inp.group(2)), string)
```
You don't need braces for a character class if there is only one character class.

ezyang replied:
Surely you don't want `detail+` here; that will match `detaildetaildetail` which cannot be correct. `detail?` is probably more correct.

iotamudelta replied:
done

ezyang commented on tools/amd_build/pyHIPIFY/hipify-python.py:
```diff
@@ -0,0 +1,946 @@
+    return output_string
+
+
+def find_paranthesis_end(input_string, start):
```
s/paranthesis/parenthesis/

iotamudelta replied:
done

ezyang commented on tools/amd_build/pyHIPIFY/hipify-python.py:
```diff
@@ -0,0 +1,946 @@
+
+
+    # Grab positional ranges of all kernel launchces
+    get_kernel_positions = [k for k in find_kernel_bounds(string)]
```
`list(find_kernel_bounds(string))` is more idiomatic.

But I don't see why you need to kill the generator here, because `get_kernel_positions` is used only once.

ezyang commented on tools/amd_build/pyHIPIFY/hipify-python.py:
```diff
@@ -0,0 +1,946 @@
+            kernel_start = string.find("<<<", kernel_end)
+
+            # Get kernel ending position (adjust end point past the >>>)
+            kernel_end = string.find(">>>", kernel_start) + 3
```
What if there is no end?

ezyang replied:
(Meta-point, you want to fail noisily and as soon as possible when you see something you don't expect; this makes debugging easier.)

iotamudelta replied:
raising exception now if no kernel end is found after a kernel start

Jorghi12 replied:
Yeah great point and beautiful catch!

ezyang commented on tools/amd_build/pyHIPIFY/hipify-python.py:
```diff
@@ -0,0 +1,946 @@
+                        return [(pos["kernel_name"]), (pos["template"]), (pos["kernel_launch"])]
+
+
+    def find_kernel_bounds(string):
```
You could do this much more concisely with a regex. The most important thing is that the inner wildcard needs to be lazy `.+?` but then you can easily extract all of the locations and inner text using a findall

ezyang commented on tools/amd_build/pyHIPIFY/hipify-python.py:
```diff
@@ -0,0 +1,946 @@
+
+        status = START
+
+        # Parse the string character by character
```
Isn't this going backwards?

Jorghi12 replied:
Starting at the kernel launches arguments, i.e. "(args..." move backwards and extract the kernel launch parameters, the template arguments, and then the kernel name. 

kernel_name<template1,template2><<<..,..,..,..>>>(args...)

ezyang commented on tools/amd_build/pyHIPIFY/hipify-python.py:
```diff
@@ -0,0 +1,946 @@
+        }
+
+        # Count for balancing template
+        count = {"<>": 0}
```
Can't you just do `count = 0`?

Jorghi12 replied:
Yeah definitely! The dictionary was just clearer for me personally.

Jorghi12 replied:
I'm in love with dictionaries.

ezyang commented on tools/amd_build/pyHIPIFY/hipify-python.py:
```diff
@@ -0,0 +1,946 @@
+
+            # Handle Kernel Name
+            if status != AT_TEMPLATE:
+                if string[i] == "(" or string[i] == ")" or string[i] == "_" or string[i].isalnum() or string[i] == ":":
```
I don't understand why you match both opening and closing parentheses here.

Jorghi12 replied:
A subset of our kernel names contain parentheses.

Jorghi12 replied:
Basically, some of our kernel names contain parentheses (and underscores).

ezyang commented on tools/amd_build/pyHIPIFY/hipify-python.py:
```diff
@@ -0,0 +1,946 @@
+                        status = AFTER_TEMPLATE
+
+            # Handle Kernel Name
+            if status != AT_TEMPLATE:
```
After all the template arguments are processed, you can use a regex to finish the rest, right?

Jorghi12 replied:
There are some cases where the arguments contain parentheses themselves. Regular expressions can't extract the text between two matching parentheses.

Jorghi12 replied:
I would but there are cases where our arguments contain parentheses themselves. Using a regular expression to capture balanced parentheses groups isn't possible.

ezyang commented on tools/amd_build/pyHIPIFY/hipify-python.py:
```diff
@@ -0,0 +1,946 @@
+
+
+def disable_asserts(input_string):
+    """ Disables regular assert statements
```
Why?

Jorghi12 replied:
There's a HIP bug that prevents asserts from compiling correctly.

Jorghi12 replied:
HIP would throw compilation errors whenever we used assert calls. Though, this may or may not have been resolved in the newly released versions.

iotamudelta replied:
This change is still valid/necessary.

ezyang commented on tools/amd_build/pyHIPIFY/hipify-python.py:
```diff
@@ -0,0 +1,946 @@
+
+
+def find_paranthesis_end(input_string, start):
+    inside_paranthesis = False
```
I don't see why you need this. Don't you have the invariant `inside_paranthesis = parans > 0`? (Also, it's parenthesis...)

ezyang commented on tools/amd_build/pyHIPIFY/hipify-python.py:
```diff
@@ -0,0 +1,946 @@
+        0 - Remove the function entirely (includes the signature).
+        1 - Stub the function and return an empty object based off the type.
+        2 - Add !defined(__HIP_PLATFORM_HCC__) preprocessors around the function.
+        3 - Add !defined(__HIP_DEVICE_COMPILE__) preprocessors around the function.
```
What's the difference between these two preprocessors? How do I decide which one?

iotamudelta replied:
`__HIP_PLATFORM_HCC__` is turned on if the compiler is `hcc`
`__HIP_DEVICE_COMPILE__` is turned on in the device path when using either `hcc` or `nvcc` through `HIP`

see: https://github.com/ROCm-Developer-Tools/HIP/blob/master/include/hip/hip_common.h

ezyang replied:
Put it in the code? :)

iotamudelta replied:
commented in enum

ezyang commented on tools/amd_build/pyHIPIFY/hipify-python.py:
```diff
@@ -0,0 +1,946 @@
+        function - Disables the function by name only.
+            e.g. "overlappingIndices"
+
+    replace_style - The style to use when stubbing functions.
```
Magic numbers here make me unhappy. Use an enum, or at the very least define constants for each of these numbers.

iotamudelta replied:
done

Jorghi12 replied:
Yeah I agree enums have much better aesthetics.

ezyang commented on tools/amd_build/pyHIPIFY/hipify-python.py:
```diff
@@ -0,0 +1,946 @@
+    return output_string
+
+
+def disable_function(input_string, function, replace_style):
```
I'd like to encourage us to stop using this function, and insert the preprocessor macros directly into the source files. We're OK with it, and it makes maintenance easier (less action at a distance>)

ezyang commented on tools/amd_build/pyHIPIFY/hipify-python.py:
```diff
@@ -0,0 +1,946 @@
+        f.seek(0)
+        f.write(contents)
+        f.truncate()
+        f.flush()
```
This (and the fsync below) is totally unnecessary, it will flush on file close.

iotamudelta replied:
done

ezyang commented:
I ran out of steam for my review. I'd suggest staging the refactors in the rocm repo first and then merging here.

iotamudelta committed:
autopep3 the pyHIPIFY scripts.

iotamudelta committed:
Include newer shared copyrith.

Update the copyright dates and reflect the reality that this has
been a collaborative effort between AMD and Facebook in the last
years.

iotamudelta committed:
Do not clone pyHIPIFY repository.

Script is now included in pytorch source.

iotamudelta committed:
Merge pull request #12 from iotamudelta/master

first round of changes to update PR

iotamudelta committed:
Merge remote-tracking branch 'upstream/master'

iotamudelta commented on tools/amd_build/build_pytorch_amd.py:
```diff
@@ -64,4 +64,4 @@
         "--output-directory", proj_dir,
         "--include-dirs"] + include_dirs + ["--yaml-settings", yaml_file, "--add-static-casts", "True"]
-os.execv("/opt/rocm/bin/hipify-python.py", ['python'] + args)
+os.execv(os.path.join(amd_build_dir, "pyHIPIFY", "hipify-python.py"), ['python'] + args)
```
done

iotamudelta committed:
Merge pull request #14 from iotamudelta/master

merge from upstream

iotamudelta committed:
Fix spelling of parenthesis.

iotamudelta committed:
Get rid of flush/fsync.

iotamudelta committed:
Fix enum error that pep8 revealed.

iotamudelta committed:
Line length fix.

iotamudelta committed:
Revert "Line length fix."

This reverts commit 864dbe416dd203b779c442d900841867960915b3.

iotamudelta committed:
Send progress info to stderr by review comment.

iotamudelta committed:
Change to detail? as per review.

iotamudelta committed:
Document what the macros stand for.

iotamudelta committed:
raise exception if no kernel end found.

iotamudelta committed:
Convert to enum from magic numbers for function disable mode.

iotamudelta committed:
Merge pull request #15 from iotamudelta/master

next round of fixes to address comments

iotamudelta commented:
@ezyang `flake8` now only complains about line length - I am not sure there is a good solution to this problem (based on also trying `pep8ify`), any suggestions?

iotamudelta committed:
Merge remote-tracking branch 'upstream/master'

ezyang commented:
Let's just exclude it for now. you can do that by editing the tox.ini

iotamudelta committed:
Merge pull request #16 from iotamudelta/master

merge from upstream

iotamudelta committed:
After discussion in review, disable flake8 on pyHIPIFY for now.

iotamudelta committed:
Merge pull request #17 from iotamudelta/master

After discussion in review, disable flake8 on pyHIPIFY for now.

iotamudelta committed:
Merge remote-tracking branch 'upstream/master'

ezyang commented on tools/amd_build/pyHIPIFY/constants.py:
```diff
@@ -0,0 +1,47 @@
+CONV_VERSION = 0,
```
This comment is still unaddressed

iotamudelta replied:
done - also referencing the hipify-clang source this came from

iotamudelta committed:
Merge pull request #18 from iotamudelta/master

Merge from pytorch upstream

ezyang commented on tools/amd_build/pyHIPIFY/hipify-python.py:
```diff
@@ -0,0 +1,967 @@
+
+def openf(filename, mode):
+    if sys.version_info[0] == 3:
+        return open(filename, mode, errors='ignore')
```
Previous comment was:

> This should not be necessary. All of the files in our repository should be UTF-8 encoded. Is this not true?

This is still not addressed.

iotamudelta replied:
Sorry, yes. There is one file @jithunnair-amd located that is not UTF-8 encoded. @jithunnair-amd could you elaborate here?

jithunnair-amd replied:
```
rocm-user@480e3f4175c2:~/pytorch__ROCM__master$ git diff --cached torch/csrc/jit/passes/shape_analysis.cpp
diff --git a/torch/csrc/jit/passes/shape_analysis.cpp b/torch/csrc/jit/passes/shape_analysis.cpp
index be18bf1..4c24c5c 100644
--- a/torch/csrc/jit/passes/shape_analysis.cpp
+++ b/torch/csrc/jit/passes/shape_analysis.cpp
@@ -325,7 +325,7 @@ void PropagateShapeOnNode(Node * node, bool insert_expands) {
    case prim::ReplaceIfUndef: {
      // If types[0] has a type, then it is not defined, and the type will
      // get set to types[0] because that will be the value propagated.
-      // If its type is<C2><A0>not<C2><A0>defined, then unification is an undefined type.
+      // If its type is not defined, then unification is an undefined type.
      SHAPE_ASSERT(types.size() == 1);
      node->output()->setType(types.at(0)->shared_from_this());
      handled = true;
```

ezyang replied:
My master doesn't show anything here? https://github.com/pytorch/pytorch/blob/ea1869244ff84b256dd1f0b4de20fb604ec92bec/torch/csrc/jit/passes/shape_analysis.cpp#L328

ezyang commented on tools/amd_build/pyHIPIFY/hipify-python.py:
```diff
@@ -0,0 +1,967 @@
+import yaml
+
+from functools import reduce
+from enum import Enum
```
enum module was added in Python 3.4, which means that your script is no longer Python 2 compatible unless you pull in a BC-compat library.

iotamudelta replied:
Sure, the enum originates from avoiding the magic numbers. So shall we revert back to them?

iotamudelta replied:
added enum34 to pip requirements

iotamudelta replied:
I've added enum34 to the requirements.

ezyang commented on tools/amd_build/pyHIPIFY/hipify-python.py:
```diff
@@ -0,0 +1,967 @@
+
+def inside_included_directories(dirpath, rootpath, include_dirs):
+    """Helper method to see if filename within included directories"""
+    return reduce(lambda result, included_directory: re.match(r'(%s)\b' % os.path.join(rootpath, included_directory), dirpath) or result, include_dirs, None)
```
Stop using '%' for interpolation; use 'format' method instead.

Additionally, if you want to put a literal string into a regex, you have to escape it correctly using `re.escape`

ezyang commented on tools/amd_build/pyHIPIFY/hipify-python.py:
```diff
@@ -0,0 +1,967 @@
+        type=str,
+        default=os.getcwd(),
+        help="The root of the project.",
+        required=True)
```
If this is required, why did you give it a default?

iotamudelta replied:
I agree. What's your preference: not requiring it or removing default?

ezyang replied:
Don't require it

iotamudelta replied:
done

ezyang commented on tools/amd_build/pyHIPIFY/hipify-python.py:
```diff
@@ -0,0 +1,967 @@
+
+    # Show detailed summary
+    if show_detailed:
+        compute_stats(stats, show_detailed)
```
You don't need the second argument right. Actually, how does this even work.

iotamudelta replied:
removed - I am also surprised.

ezyang commented on tools/amd_build/pyHIPIFY/hipify-python.py:
```diff
@@ -0,0 +1,967 @@
+        nargs='+',
+        default=["cu", "cuh", "c", "cpp", "h", "in", "hpp"],
+        help="The extensions for files to run the Hipify script over.",
+        required=False)
```
I'm not sure in what situations I'd actually want to override this.

iotamudelta replied:
The only scenario I can come up with is development purposes. So shall we remove the option?

ezyang commented on tools/amd_build/pyHIPIFY/hipify-python.py:
```diff
@@ -0,0 +1,967 @@
+    parser.add_argument(
+        '--extensions',
+        nargs='+',
+        default=["cu", "cuh", "c", "cpp", "h", "in", "hpp"],
```
Don't have to do in this PR, but `in` is a bit of a dangerous extension to match over, since ANY file that is a template, e.g., `myrandom.txt.in`. It would be better if you could say `.h.in` and then adjust the code that reads the extensions to be able to handle extensions that have multiple components (if it doesn't support that already).

iotamudelta replied:
Yes, let's do that later.

ezyang commented on tools/amd_build/pyHIPIFY/hipify-python.py:
```diff
@@ -0,0 +1,967 @@
+        '--add-static-casts',
+        type=bool,
+        default=False,
+        help="Whether to automatically add static_casts to kernel arguments.",
```
Under what circumstances do I want this option?

iotamudelta replied:
Our kernel execution does not like if kernel arguments do not match their definitions. Hence why we provide this option. Since there are quite a few instances of this in PyTorch, we enable it. We can default on it, if that's preferable?

ezyang reviewed (commented):
not done yet but I want to stream these out instead of accumulating

ezyang commented on tools/amd_build/pyHIPIFY/hipify-python.py:
```diff
@@ -0,0 +1,967 @@
+
+    # If no output directory, provide a default one.
+    if args.output_directory is "":
+        args.project_directory = args.project_directory[0:-
```
args.project_directory.rstrip("/") is a more idiomatic way of spelling this.

iotamudelta replied:
done

ezyang commented on tools/amd_build/pyHIPIFY/hipify-python.py:
```diff
@@ -0,0 +1,967 @@
+    # Verify the project directory exists.
+    if not os.path.exists(args.project_directory):
+        print("The project folder specified does not exist.")
+        return
```
Don't `return` here, because that will cause your Python script to exit with exit code 0, indicating success. `sys.exit(1)` is good. There are a bunch of other returns which need to be adjusted too.

iotamudelta replied:
done

ezyang commented on tools/amd_build/pyHIPIFY/hipify-python.py:
```diff
@@ -0,0 +1,967 @@
+
+    # Make sure output directory exists.
+    if not os.path.exists(args.output_directory):
+        print("The output folder already exists.")
```
So, is the intended semantics that if a user explicitly specifies an output directory, we bail if it already exists, but if a user doesn't, then we're willing to clobber `_amd`? That seems inconsistent.

iotamudelta replied:
Agreed. We should probably always bail out.

iotamudelta replied:
done - always bailing out

ezyang commented on tools/amd_build/pyHIPIFY/hipify-python.py:
```diff
@@ -0,0 +1,967 @@
+    if not os.path.exists(args.output_directory):
+        shutil.copytree(args.project_directory, args.output_directory)
+
+    # Extract all of the kernel parameter and template type information.
```
This comment seems inconsistent with the conditional that comes below it.

ezyang replied:
Perhaps it wanted to also add, "if necessary"? Why does this have to be done now, as opposed to later?

iotamudelta replied:
I think the comment does reflect what the code is doing: if we enable the "add static_cast" feature, the script is extracting template and kernel parameters. Do you disagree?

iotamudelta replied:
concerning the if necessary: the current logic will add static_casts for all arguments, even the ones that match the type definition (script is easier that way). Concerning the now and not later: it'll first build the kernel arguments up and then does the static_cast addition in a second pass.

ezyang commented on tools/amd_build/pyHIPIFY/hipify-python.py:
```diff
@@ -0,0 +1,967 @@
+
+    # Open YAML file with disable information.
+    if args.yaml_settings != "":
+        with openf(args.yaml_settings, "r") as lines:
```
`lines` is a misnomer. Call it `f`.

iotamudelta replied:
done

iotamudelta committed:
Name f, not lines.

iotamudelta committed:
Signal error with sys.exit()

iotamudelta committed:
Always bail out if the output directory already exists.

This will hopefully safe some grief in the future with overriding code.

iotamudelta committed:
Strip trailing slash.

jithunnair-amd committed:
Fix static_cast logic in hipify script for better coverage. Remove obsolete code for output directory not existing

iotamudelta committed:
Merge pull request #19 from iotamudelta/master

Address more review comments

iotamudelta committed:
Remove second argument that wasn't actually required.

ezyang commented on tools/amd_build/pyHIPIFY/hipify-python.py:
```diff
@@ -0,0 +1,967 @@
+            if "functions" in disable_info:
+                functions = disable_info["functions"]
+            else:
+                functions = []
```
A more idiomatic way to write this is:

```
functions = disable_info.get("functions", [])
```

ezyang replied:
Ditto below.

iotamudelta replied:
done

ezyang commented on tools/amd_build/pyHIPIFY/hipify-python.py:
```diff
@@ -0,0 +1,967 @@
+                f.seek(0)
+                f.write(txt)
+                f.truncate()
+                f.close()
```
You don't have to close the file; the context manager will do it for you.

iotamudelta replied:
done

ezyang commented on tools/amd_build/pyHIPIFY/hipify-python.py:
```diff
@@ -0,0 +1,966 @@
+        # Disable unsupported HIP functions
+        for disable in yaml_data["disable_unsupported_hip_calls"]:
+            filepath = os.path.join(args.output_directory, disable["path"])
+            if "functions" in disable:
```
A more idiomatic way to spell this is

```
functions = disable.get("functions", [])
```

ditto below

iotamudelta replied:
done

ezyang commented on tools/amd_build/pyHIPIFY/hipify-python.py:
```diff
@@ -0,0 +1,966 @@
+                s_constants = []
+
+            if not os.path.exists(filepath):
+                print("\n" + bcolors.WARNING + "YAML Warning: File %s does not exist." % filepath + bcolors.ENDC)
```
Shouldn't this be a hard error?

iotamudelta replied:
No - in an ideal world no functions should need to be disabled. Or?

ezyang commented on tools/amd_build/pyHIPIFY/hipify-python.py:
```diff
@@ -0,0 +1,966 @@
+
+                # Disable Constants w\ Boundary.
+                for const in constants:
+                    txt = re.sub(r"\b%s\b" % const, constants[const], txt)
```
This is also committing the regex mistake.

ezyang commented on tools/amd_build/pyHIPIFY/hipify-python.py:
```diff
@@ -0,0 +1,966 @@
+                f.seek(0)
+                f.write(txt)
+                f.truncate()
+                f.close()
```
unnecessary

iotamudelta replied:
done

ezyang commented on tools/amd_build/pyHIPIFY/hipify-python.py:
```diff
@@ -0,0 +1,966 @@
+
+
+def add_static_casts(directory, extensions, KernelTemplateParams):
+    """Added necessary static casts to kernel launches due to issue in HIP"""
```
A great way to make this documentation more clear is to give some example code before and after this transformation.

Also, if you cite an issue in HIP, best to cite the actual issue (e.g. a URL), so that it's easy to see if the problem in HIP has been fixed or not.

jithunnair-amd replied:
It's actually not an *issue* in HIP. Rather, it's to ensure that the type of the kernel arguments matches the type specified in the kernel definition. I have updated the comments and added a before-and-after example in commit commit acfacbc as part of PR https://github.com/ROCmSoftwarePlatform/pytorch/pull/20

ezyang commented on tools/amd_build/pyHIPIFY/hipify-python.py:
```diff
@@ -0,0 +1,966 @@
+                with openf(filepath, "r+") as fileobj:
+                    input_source = fileobj.read()
+                    new_output_source = input_source
+                    get_kernel_definitions = [k for k in re.finditer("hipLaunchKernelGGL\(", input_source)]
```
The generator is used only once, you don't have to listify it. To make it less error prone, move the generator inside the for-loop (don't bind a name to it)

iotamudelta replied:
Changed this here and in another place. There is still more but those have such complicated matches that I find the current code more readable.

ezyang commented on tools/amd_build/pyHIPIFY/hipify-python.py:
```diff
@@ -0,0 +1,966 @@
+        f.truncate()
+
+
+def extract_arguments(start, string):
```
So, IIUC, this requires you to pass it a string that starts with a parentheses, right? That's a very important invariant that should be documented here.

jithunnair-amd replied:
The string doesn't have to *start* with a parenthesis, but does need to have the arguments enclosed within parentheses as part of the string

jithunnair-amd replied:
Function requirement clarified in comments with an example in commit acfacbc as part of PR https://github.com/ROCmSoftwarePlatform/pytorch/pull/20

iotamudelta committed:
Don't require the root of the project to be specified.

We already had a fallback.

iotamudelta committed:
Python-ify as per review.

iotamudelta committed:
Do not close file explicitly.

iotamudelta committed:
Python-ify as per review.

jithunnair-amd committed:
Replace old string replacement technique with new technique using .format()

jithunnair-amd committed:
Merge branch 'master' into hipify_changes

jithunnair-amd committed:
Add comments for add_static_cast and extract_arguments functions

iotamudelta committed:
Merge pull request #21 from iotamudelta/master

Address even more comments

iotamudelta committed:
Merge pull request #20 from jithunnair-amd/hipify_changes

Fix static_cast logic in hipify script for better coverage

iotamudelta committed:
Add enum34 as a requirement for backwards compatibility.

iotamudelta committed:
Merge pull request #23 from iotamudelta/master

Add enum34 as a requirement for backwards compatibility.

iotamudelta committed:
Merge remote-tracking branch 'upstream/master'

iotamudelta committed:
Merge pull request #24 from iotamudelta/master

Merge from upstream

iotamudelta committed:
Merge remote-tracking branch 'upstream/master'

iotamudelta committed:
Merge pull request #25 from iotamudelta/master

Merge from upstream

iotamudelta committed:
Use generator directly in loop as by review

iotamudelta committed:
Merge pull request #26 from iotamudelta/master

Use generator directly in loop as by review

Commit:
Escape input strings

String inputs should have meta characters escaped prior to using them with regular expressions.

iotamudelta committed:
Revert "Escape input strings". No direct commits to this repository.

This reverts commit 4158af1d403ed08d8e2deb5f4ce306fcec563a64.

Commit:
Escape inputs into reg. ex.

The inputs into the regular expressions must have their meta characters escaped or unexpected replacements may occur.

iotamudelta committed:
Revert "Escape inputs into reg. ex." Still no direct commits to master.

This reverts commit cf2c228c45a1ed2ac8dc6b51482eac34590d743d.

Commit:
test

test

iotamudelta committed:
Revert "test". No direct commits.

This reverts commit 08043e3ad49f8705f06dcc469a06ce47319ee0f7.
